src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -491,109 +499,110 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
 
                     var state = ReadState();
 
-                    // if we already have a compilation, we must be already done!  This can happen if two
-                    // threads were waiting to build, and we came in after the other succeeded.
-                    if (state is FinalState finalState)
-                        return finalState;
-
-                    var compilationWithoutGeneratedDocuments = state.CompilationWithoutGeneratedDocuments;
-                    if (compilationWithoutGeneratedDocuments == null)
+                    return state switch
```

Comment:
this if/else chain becomes a much cleaner switch that shows that it handles all cases.